lib/web_ui/lib/src/engine/text_editing/text_editing.dart

```diff
@@ -440,6 +440,138 @@ class AutofillInfo {
   }
 }
 
+/// Replaces a range of text in the original string with the text given in the
+/// replacement string.
+String _replace(String originalText, String replacementText, ui.TextRange replacedRange) {
+  assert(replacedRange.isValid);
+  assert(replacedRange.start <= originalText.length && replacedRange.end <= originalText.length);
+
+  final ui.TextRange normalizedRange = ui.TextRange(start: math.min(replacedRange.start, replacedRange.end), end: math.max(replacedRange.start, replacedRange.end));
+
+  final String newText = normalizedRange.textBefore(originalText) + replacementText + normalizedRange.textAfter(originalText);
+
+  return newText;
+}
+
+class TextEditingDeltaState {
+  TextEditingDeltaState({
+    this.oldText = '',
+    this.deltaText = '',
+    this.deltaStart = -1,
+    this.deltaEnd = -1,
+    this.baseOffset = -1,
+    this.extentOffset = -1,
+    this.composingOffset = -1,
+    this.composingExtent = -1,
+  });
+
+  static TextEditingDeltaState inferDeltaState(EditingState newEditingState, EditingState? lastEditingState, TextEditingDeltaState lastTextEditingDeltaState) {
+    final bool previousSelectionWasCollapsed = lastEditingState?.baseOffset == lastEditingState?.extentOffset;
+
+    if (lastTextEditingDeltaState.deltaText.isEmpty && lastTextEditingDeltaState.deltaEnd != -1) {
+      // We are removing text.
+      // When text is deleted outside of the composing region or is cut using the native toolbar,
+      // we calculate the length of the deleted text by comparing the new and old editing state lengths.
+      // This value is then subtracted from the end position of the delta to capture the deleted range.
+      final int deletedLength = lastTextEditingDeltaState.oldText.length - newEditingState.text!.length;
+      lastTextEditingDeltaState.deltaStart = lastTextEditingDeltaState.deltaEnd - deletedLength;
```

Comment:
Related to a comment I'm leaving below where inferDeltaState is called: What do you think of not modifying the lastTextEditingDelta argument?  Maybe just make a copy if that makes sense.

I'm not sure if this is in the Flutter style guide, but I personally try not to modify function arguments unless I have to.  Seeing this makes me think that the variable passed in as lastTextEditingDelta where inferDeltaState is called is expecting to be modified and will be reused for something.

Reply:
I see what you mean. I think it makes sense to not modify the lastTextEditingDelta argument since we are constructing a new delta with the new information we have at this point in the flow.